### PR TITLE
Rewrite hb_report use python 

### PR DESCRIPTION
**Pull request by wenshizhang**
Hb_report is part of crmsh, it would be better if hb_report implement in
python like crmsh.

For now, hb_report can do most works:
 - parse parameters user input
 - get environment variables
   - cluster node list
   - from time and to time
   - and so on
 - find logs and configure files location
 - create environment variable xmlfile send it to cluster nodes
 - cluster node collect logs and send it back
 - remove tempfiles

Here are several things need to do in future:
 - analyze logs:
   - check which logs we need do not exists
   - check if logs from different node are different
   - check other stuff
 - check if log empty
 - CTS stuff
 - heartbeat cluster support configure(ha_support_conf.py)

There are few classes, please read README.md. Because of unfinished
features and not full tested, there are two hb_report(hb_report
	implement in shell and python_hb_report implement in python) in
this repo.

Signed-by-off: Shiwen Zhang <szhang@suse.com>

### CODE REVIEW

File: python_hb_report/corosync_conf_support.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		return output
+
+def essential_files():
+	file_info = []
+	file_info.append(['d',envir.PCMK_LIB,'0750','hacluster','haclient'])
+	file_info.append(['d',envir.PE_STATE_DIR,'0750','hacluster','haclient'])
+	file_info.append(['d',envir.CIB_DIR,'0750','hacluster','haclient'])
+	
+	return file_info
+
```
Comment (liu4480) [inline python_hb_report/corosync_conf_support.py]: There are too many blank lines, could you remove them?

File: python_hb_report/collector.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+		msg = msg +'\n'+ scsi_info
+
+		pci_info = utillib.do_command(['lspci'])
+		msg = msg +'\n'+ pci_info
+
+		mount_info = utillib.do_command(['mount'])
+		msg = msg +'\n' + mount_info
+
+		#df can block, run in background, allow for 5 seconds
+		df_pro = subprocess.Popen(['df'],stderr = subprocess.STDOUT,stdout = subprocess.PIPE)
```
Comment (liu4480) [inline python_hb_report/collector.py]: why not use utillib.do_command here?

File: python_hb_report/collector.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+
+	def dlm_dump(self):
+		output = os.path.join(self.WORKDIR,envir.DLM_DUMP_F)
+		if not utillib.do_which('dlm_tool'):
+			return 
+		dlm_file_info = 'NOTICE - Lockspace overview:\n'	
+		dlm_file_info = dlm_file_info + utillib.do_command(['dlm_tool','ls'])
+
+		dlm_pro = subprocess.Popen(['dlm_tool','ls'],stdout = subprocess.PIPE,stderr = subprocess.STDOUT)
+		grep_pro = subprocess.Popen(['grep','name'],stdin = dlm_pro.stdout,stdout = subprocess.PIPE,stderr = subprocess.STDOUT)
```
Comment (liu4480) [inline python_hb_report/collector.py]: in line 145, you 've called dlm_tool ls, and in line 147, you call it again, why?

File: python_hb_report/collector.py [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+			for f in flist:
+				if not self.skip_lvl(1):
+					path = os.path.join(workdir,utillib.basename(envir.PE_STATE_DIR))
+					path = os.path.join(path,utillib.basename(f))
+					self.pe2dot(path)
+		else:
+			utillib.debug('too many PE inputs to create dot files')
+
+	def touch_DC_if_dc(self):
+		dc = utillib.do_command(['crmadmin','-D'])
```
Comment (liu4480) [inline python_hb_report/collector.py]: sometimes utillib.do_command, and sometimes subprocess.Popen, is there any reason to do so?

File: python_hb_report/collector.py [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+			dirs = os.path.join(self.WORKDIR,envir.CIB_F).split()
+		except OSError:
+			pass
+		else:
+			need_replace_files.extend(dirs)
+
+		try:
+			dirs = os.listdir(os.path.join(self.WORKDIR,'pengine'))
+		except OSError:
+			pass
```
Comment (liu4480) [inline python_hb_report/collector.py]: Doesn't here needs warning message?

File: python_hb_report/node.py [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+		'''
+
+		#first get CORE_DIRS and PACKAGES
+		if envir.HA_VARLIB != envir.PCMK_LIB:
+			envir.CORES_DIRS.append(envir.HA_VARLIB+"/cores")
+			envir.CORES_DIRS.append(envir.PCMK_LIB+'/cores')
+		else:
+			envir.CORES_DIRS.append(envir.HA_VARLIB+'/cores')
+
+		packages = 'pacemaker libpacemaker3 pacemaker-pygui pacemaker-pymgmt pymgmt-client openais libopenais2 libopenais3 corosync libcorosync4 resource-agents cluster-glue libglue2 ldirectord libqb0 heartbeat heartbeat-common heartbeat-resources libheartbeat2 booth ocfs2-tools ocfs2-tools-o2cb ocfs2console ocfs2-kmp-default ocfs2-kmp-pae ocfs2-kmp-xen ocfs2-kmp-debug ocfs2-kmp-trace drbd drbd-kmp-xen drbd-kmp-pae drbd-kmp-default drbd-kmp-debug drbd-kmp-trace drbd-heartbeat drbd-pacemaker drbd-utils drbd-bash-completion drbd-xen lvm2 lvm2-clvm cmirrord libdlm libdlm2 libdlm3 hawk ruby lighttpd kernel-default kernel-pae kernel-xen glibc'
```
Comment (liu4480) [inline python_hb_report/node.py]: it is better not to have such a long line here.

File: python_hb_report/utillib.py [hunk 7]
```diff
@@ -0,0 +1,924 @@
...
+	date = datetime.datetime.date(now).strftime("%a-%d-%m-%Y")
+	envir.DEST = "hb_report-"+date
+	envir.SANITIZE.append("passw.*")
+	envir.SSH_OPTS = ['-o StrictHostKeyChecking=no','-o EscapeChar=none','-o ConnectTimeout=15']
+	envir.LOG_PATTERNS.append("CRIT:")
+	envir.LOG_PATTERNS.append("ERROR:")
+
+def debug(msg):
+	if (envir.VERBOSITY >0):
+		print >> sys.stderr,socket.gethostname(),"DEBUG:",msg
```
Comment (liu4480) [inline python_hb_report/utillib.py]: also in line 44,48,51

File: python_hb_report/utillib.py [hunk 8]
```diff
@@ -0,0 +1,924 @@
...
+	return dirname
+
+def basename(path):
+	if path.endswith("/"):
+		path = path[1:len(path)-1]
+	
+	index = path.rfind("/")
+	basename = path[index+1:]
+	
+	return basename
```
Comment (liu4480) [inline python_hb_report/utillib.py]: os.path.basename can do this job

File: python_hb_report/utillib.py [hunk 9]
```diff
@@ -0,0 +1,924 @@
...
+		if re.match("\d+",d):
+			path = os.path.join('/proc',d+'/cmdline')
+			f = open(path,'r')
+			msg = f.readline()
+			if msg.find(proname) != -1 and msg.find('grep') == -1:
+				return False
+	return True
+
+def findmsg(mark):
+	syslog = '/var/log /var/logs /var/syslog /var/adm /var/log/ha /var/log/cluster /var/log/pacemaker /var/log/heartbeat /var/log/crm /var/log/corosync'
```
Comment (liu4480) [inline python_hb_report/utillib.py]: it is better to use """ """ to change this long line to multiple short lines

File: python_hb_report/utillib.py [hunk 10]
```diff
@@ -0,0 +1,924 @@
...
+
+	return tmp
+
+def add_tmpfiles(files):
+	
+	if not os.path.isfile(envir.__TMPFLIST):
+		return
+	f = open(envir.__TMPFLIST,'a')
+	f.write(files)
+	f.close
```
Comment (liu4480) [inline python_hb_report/utillib.py]: lose () after close

File: python_hb_report/utillib.py [hunk 11]
```diff
@@ -0,0 +1,924 @@
...
+
+def getstamp_syslog(message):
+	return ''.join(message.split()[0:2])
+
+def find_getstampproc(sla,filepath):
+	'''
+	Now  we do not need to get log from /var/log/messages and /var/log/pacemaker.log
+	'''
+	pass
+
```
Comment (liu4480) [inline python_hb_report/utillib.py]: blank lines

File: python_hb_report/node.py [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+						envir.CRM_DAEMON_DIR = p+'/'+d+'/'+d2+'/crmd'
+						break
+	
+	def collect_journal(self,workdir):
+		'''
+		Collect Journal from Systemd, then write the result to file journal.log 
+		'''
+		global outf
+		from_time = str(int(envir.FROM_TIME))
+		to_time = str(int(envir.TO_TIME))
```
Comment (nick-wang) [inline python_hb_report/node.py]: why not directly use str(envir.FROM_TIME)

File: python_hb_report/README.md [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+##master.py
+
+The defination of master class, when user use the hb_report command, in the crm report script, some secure things is done, then crm script call master_node run function. Master do all things  master node should do, like decide to which log is collected, check ssh connections and so on.
+### class master
+####MEMBER VARABLES
+* PIDS: master create new processes to collect logs, master should wait those processes exit. PIDS stores those processes' pid
+####MEMBER FUNCTIONS
+* version: print hb_report version information
+* usage: print help message
+* analyzed_argvment: analyze parameters user input
```
Comment (liu4480) [inline python_hb_report/README.md]: argvment should be argument

File: python_hb_report/utillib.py [hunk 14]
```diff
@@ -0,0 +1,924 @@
...
+import tempfile
+
+import xml.etree.ElementTree as ET
+from xml.dom import minidom
+
+from crmsh	import utils
+
+
+#set variables default
+def setvarsanddefaults():
```
Comment (nick-wang) [inline python_hb_report/utillib.py]: Better to use _ to split words.

File: python_hb_report/README.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+The defination of node class, it is the father class of master and slave class, abstract some common features about the subclass.
+###class node
+##envir.py
+This is environment variable configure file, all the  variable is needed during the hb_report run.Kind of important
+###ENVIRONMENT VERABLES
+##utillib.py
+This is functions library.
+###FUNCTIONS
+##corosync_conf_support.py
+This is the corosync cluster support script, this script implement some unique function for corosync cluster
```
Comment (liu4480) [inline python_hb_report/README.md]: implements should be implements

File: python_hb_report/envir.py [hunk 16]
```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/python3.5
```
Comment (nick-wang) [inline python_hb_report/envir.py]: Why use python3.5? SLE12SP2 still using python2.7 as default.

File: python_hb_report/utillib.py [hunk 17]
```diff
@@ -0,0 +1,924 @@
...
+	envir.LOG_PATTERNS.append("CRIT:")
+	envir.LOG_PATTERNS.append("ERROR:")
+
+def debug(msg):
+	if (envir.VERBOSITY >0):
+		print >> sys.stderr,socket.gethostname(),"DEBUG:",msg
+		return 0
+
+def fatal(msg):
+	print >> sys.stderr,socket.gethostname(),"ERROR:",msg
```
Comment (nick-wang) [inline python_hb_report/utillib.py]: Not sure python3 support "print >> sys.stderr,socket.gethostname(),"ERROR:",msg" or not?

File: python_hb_report/collector.py [hunk 18]
```diff
@@ -0,0 +1,374 @@
...
+
+
+
+
+	def getpeinputs(self,workdir):
+		i = 0
+
+		utillib.debug('looking for PE files in'+envir.PE_STATE_DIR)
+		flist = utillib.find_files(envir.PE_STATE_DIR.split())
+		grep_pro = subprocess.Popen(['grep','-v',"[.]last$"],stdin = subprocess.PIPE,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)
```
Comment (liu4480) [inline python_hb_report/collector.py]: why not use utillib.do_command here, as utillib.do_command is wrapper of subprocess.Popen, and you just get the output, does have extra operation.

File: python_hb_report/collector.py [hunk 19]
```diff
@@ -0,0 +1,374 @@
...
+		flist = grep_pro.communicate(' '.join(flist))[0].split()
+		
+		if len(flist):
+			filename = utillib.basename(envir.PE_STATE_DIR)
+			pengine_dir = os.path.join(workdir,filename)
+			os.mkdir(pengine_dir)
+			for f in flist:
+#				os.symlink(f,os.path.join(pengine_dir,utillib.basename(f)))
+				os.symlink(f,pengine_dir)
+				utillib.do_command(['ln','-s',f,pengine_dir])
```
Comment (liu4480) [inline python_hb_report/collector.py]: what is difference between line 180 and line 181

File: python_hb_report/utillib.py [hunk 20]
```diff
@@ -0,0 +1,924 @@
...
+	return value
+
+def dirname(path):
+	if path.endswith("/"):
+		path = path[1:len(path)-1]
+
+	index = path.rfind("/")
+	dirname = path[0:index]
+	
+	return dirname
```
Comment (nick-wang) [inline python_hb_report/utillib.py]: os.path.dirname() can do this job

File: python_hb_report/collector.py [hunk 21]
```diff
@@ -0,0 +1,374 @@
...
+
+	def getratraces(self):
+		i = 0
+		trace_dir = os.path.join(envir.HA_VARLIB,'trace_ra')
+		
+		if not os.path.isdir(trace_dir):
+			return False
+
+		utillib.debug('looking for RA trace files in '+trace_dir)
+		sed_pro = subprocess.Popen(['sed',"s,"+utillib.dirname(trace_dir)+"/,,g"],stdin = subprocess.PIPE,stdout = subprocess.PIPE)
```
Comment (liu4480) [inline python_hb_report/collector.py]: it seems utillib.do_command can do the job here, as it is wrapper of subprocess.Popen, and here has no extra operation.

File: python_hb_report/utillib.py [hunk 22]
```diff
@@ -0,0 +1,924 @@
...
+def logd_getcfvar(pattern):
+	'''
+	TODO
+	'''
+	f = open(envir.LOGD_CF)
+	for line in f:
+		if line.startswith('#'):
+			continue
+		if line.startwith(pattern):
+			pass
```
Comment (nick-wang) [inline python_hb_report/utillib.py]: What's the purpose of this function?

File: python_hb_report/utillib.py [hunk 23]
```diff
@@ -0,0 +1,924 @@
...
+	
+	if not len(log):
+		debug('no HA log found in '+dirs)
+	else:
+		debug('found HA log at'+' '.join(log))
+	
+	return log
+
+
+def finf_getstampproc():
```
Comment (nick-wang) [inline python_hb_report/utillib.py]: Is that find_getstampproc?  If so, there is another dummy function named find_getstampproc

File: python_hb_report/utillib.py [hunk 24]
```diff
@@ -0,0 +1,924 @@
...
+	Test whether crm is running
+	if running return True, otherwise return False
+	'''
+	result = 0
+	#if ps and grep find the crmd then return True
+	if  not ps_grep('crmd'):
+		return True
+	pid = os.fork()
+	if not pid:
+		result = os.system('crmadmin -D >/dev/null 2>&1')
```
Comment (liu4480) [inline python_hb_report/utillib.py]: subprocess.Popen has a returncode, you can also use it.

File: python_hb_report/utillib.py [hunk 25]
```diff
@@ -0,0 +1,924 @@
...
+	fd.close()
+
+	return ''
+		
+
+def findmsg():
+	'''
+	Found HA Log
+	'''
+	dirs="/var/log /var/logs /var/syslog /var/adm /var/log/ha /var/log/cluster /var/log/pacemaker /var/log/heartbeat /var/log/crm /var/log/corosync /var/log/openais"
```
Comment (liu4480) [inline python_hb_report/utillib.py]: too long line

File: python_hb_report/utillib.py [hunk 26]
```diff
@@ -0,0 +1,924 @@
...
+
+	if not do_which(command):
+		debug(command+' is not found')
+		msg = command+' : command not found'
+		return msg
+
+	com_pro = subprocess.Popen(comm_list,stdout = subprocess.PIPE,stderr = subprocess.STDOUT)
+
+	msg = com_pro.communicate()[0]
+	return msg
```
Comment (liu4480) [inline python_hb_report/utillib.py]: This function can return both the output and return value.

File: python_hb_report/utillib.py (hunk 17 above)
Comment (liu4480) [inline python_hb_report/utillib.py]: here is what python3 defined for print:
print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)

File: python_hb_report/utillib.py (hunk 26 above)
Comment (liu4480) [inline python_hb_report/utillib.py]: This function can return both the output and return value of child process.

**Comment by krig** [comment]
OK, looks like there are plenty of reviews to read already :)

Don't worry about the failing Travis CI check, it is unrelated. I need to fix that.

File: python_hb_report/collector.py [hunk 27]
```diff
@@ -266,9 +226,11 @@ def sanitize(self):
 		Replace sensitive info with ****
 		'''
 		need_replace_files = []
-		for f in os.path.join(self.WORKDIR,envir.B_CONF).split():
-			if os.path.isfile(f):
-				utillib.sanitize_one(f)
+		for b in envir.B_CONF.split():
+			print b
```
Comment (liu4480) [inline python_hb_report/collector.py]: why do you print b here? For debugging?

File: python_hb_report/README.md [hunk 28]
```diff
@@ -0,0 +1,60 @@
+#Descriptions
+This directory include hb_report all the codes(implements with python). There are three classed, one environment variable configure and one lib. The following documents explain these files.
+
+#Explainations
+##master.py
+
+The defination of master class, when user use the hb_report command, in the crm report script, some secure things is done, then crm script call master_node run function. Master do all things  master node should do, like decide to which log is collected, check ssh connections and so on.
```
Comment (liu4480) [inline python_hb_report/README.md]: some secure things are done

File: python_hb_report/README.md [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+* event:
+* check_if_log_empty: master check logs that is collected is empty
+* final_word: print final work
+* send_env: master get environment and logs name need to be collected, create xmlfile, then call send_env scp xmlfile to collector /tmp.
+* get_user_node_cts: TODO
+* get_cts_log: TODO
+* is_member: find user node is cluster member or not
+
+##collector.py
+The defination of collector class, this is a simple class, do collect log things and send the resule to master node.
```
Comment (liu4480) [inline python_hb_report/README.md]: resule? result?

File: python_hb_report/node.py [hunk 30]
```diff
@@ -0,0 +1,392 @@
+#!/usr/bin/python3.5
```
Comment (liu4480) [inline python_hb_report/node.py]: there are still python3.5 here.

**Comment by krig** [comment]
Hi Shiwen, reviewers: I have pushed this pull request as a branch named `hb_report.py`:

https://github.com/ClusterLabs/crmsh/tree/hb_report.py

This way, we can continue the work on it without having to make it "perfect" in one single pull request.

Thank you!

File: python_hb_report/master.py [hunk 31]
```diff
@@ -348,8 +345,30 @@ def start_slave_collector(self,nodes,port=22,username='root'):
 		fderr.write(errmsg)
 		fderr.close()
 
+	def events_all(self,logf,outf):
+		epatt= []
+		logfd = open(logf,'r')
+		lines = logfd.readlines()
+		print lines
```
Comment (liu4480) [inline python_hb_report/master.py]: what is the purpose to print here? for debugging?

**Comment by krig** [comment]
Hi @wenshizhang, I will close this pull request in favor of #195 which completes the rewrite, thank you!

**Pull request closed without merging**